packages/xo-server/src/index.js

```diff
@@ -140,6 +148,54 @@ async function setUpPassport(express, xo) {
     res.redirect('/')
   })
 
+  express.get('/signin/otp', (req, res, next) => {
+    const { user } = req.session
+
+    if (!user) {
+      return res.redirect('/signin')
+    }
+
+    res.send(
+      signInPage({
+        error: req.flash('error')[0],
+        otp: true,
+        strategies,
+      })
+    )
+  })
+
+  express.post('/signin/otp', (req, res, next) => {
+    const { body, session } = req
+    const { user } = session
+
+    if (!user) {
+      return res.redirect('/signin')
+    }
+
+    const isValid = authenticator.check(body.otp, user?.preferences?.otp)
```

Comment:
`user.preferences.otp`: we know `user` and `user.preferences.otp` are defined, otherwise we would not be there.